designs/aws-launch-template-network-interfaces.md

```diff
@@ -0,0 +1,69 @@
+# AWS LaunchTemplate's NetworkInterfaces
+AWS launchTemplates object enables the [configuration of network interfaces](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-networkinterface.html). 
+This gives the users flexibility they are currently lacking because Karpenter does not expose networkInterfaces in the `AWSNodeTemplate` API. here are some of the usecases requested:
+
+* #2026 Associate a public IPv4 address to the network interface
+* #3369 Create a node with multiple network interfaces
+* #3369 create a node with EFA(Elastic Fabric Adapter) network interface
+
+# challenges:
+exposing networkInterfaces as part of the `AWSNodeTemplate` and using the data to configure AWS LaunchTemplate would solve the problem for the users. However, there are some challenges that we need to address:   
+
+## subnet and security group resolution logic
+currently security groups and subnets are part of the `AWSNodeTemplate`; and they are resolved from tags at run-time. This implies the resolved subnet and security group will be the same for all network interfaces.
+this is counterintuitive as users expect to be able to configure different subnets and security groups for each network interface.
+```yaml
+apiVersion: karpenter.k8s.aws/v1alpha1
+kind: AWSNodeTemplate
+spec:
+  subnetSelector:
+      label: subnet-1 # users expect this to be used for all network interfaces
+  securityGroupSelector:
+      securityGroup: sg-1 # users expect this to be used for all network interfaces
+  networkInterfaces:
+  - subnetSelector:
+        subnet: subnet-2 # users expect this to be used for this specific network interface
+    securityGroupSelector:
+        securityGroup: sg-2 # users expect this to be used for this specific network interface
+```
+having both options (`AWSNodeTemplate` level and networkInterface level) will complicate the API and will cause confusion for the users. It shall also complicate the logic on karpenter controller side.   
+In addition, removing the `AWSNodeTemplate`'s level subnet and security group selectors is a breaking change from API backward compatibility perspective.
+
+one more thing to consider is that subnets are bound to a single AZ. that's why the resolution happen in runtime instead of being part of the AWS launchTemplate (to avoid creating as many launch templates as the AZs in the zone).
+moving the resolution logic to the networkInterface level has the following complications: 
+1. more resolutions and interface allocations to do
+2. NetworkInterfaces can have conflicting requirements (e.g. different subnets in different AZs). which we need to handle in runtime.
+
+## networkInterfaces with specific IP addresses
```

Comment:
Agree on leaving this out!

I'm not entirely sure how this would work, but there is an ask for a selector on interfaceID such that you could assign an EIP, and could extend that to a NetworkInterfaceID https://github.com/aws/karpenter/issues/3899 

I'm not sure if Karpenter is the right place for that logic though, seems more like the CNI should allocate the EIP rather than Karpenter.

Reply:
There is no way to assign an EIP to launchTemplate. The best we can do is 

1. use a selector to select a NetworkInterface and assign it to the launchTemplate using [NetworkInterfaceId](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-networkinterface.html#cfn-ec2-launchtemplate-networkinterface-networkinterfaceid); in this case the user will have to create the network interfaces and assign the EIP to them beforehand. Karpenter do not allocate EIPs in that case
2. alternatively, regardless of how the networkInterfaces are created (dynamically, or using a networkInterface selector to pick an existing interface); users can specify an EIP tag/selector that karpenter will use it to associate the EIP to the EC2 instance after it is created.

I guess 1 is more in the scope of this change, but it does not fully solve the problem for the user. they will need to pre-create network interfaces and associate EIPs with them.

in addition, the [workaround](https://github.com/aws-samples/eip-and-dns-auto-binding-for-karpenter-worker#solution-introduction) specified by the user in the above issue suggests more than just assigning an EIP which go beyond the scope of Karpenter